compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -732,6 +734,10 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         match self.solver_mode() {
             SolverMode::Normal => return,
             SolverMode::Coherence => {
+                if goal.predicate.polarity() == ty::ImplPolarity::Negative {
+                    return;
+                }
```

Comment:
why is this correct? it feels dubious, can you add a comment

Reply:
That's what the old solver does. I don't actually *know* if it's correct, but I'll need to think about it for a while.

Reply:
When proving `&str: !Error`, we are assembling two candidates: the unknowable one, and the one from [the impl](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#impl-Error-for-%26str). The impl actually returns two trivial region constraints (consequence of the way that region equality works when matching impl), so we don't end up choosing a candidate when merging --> ambiguity.

We need to do that thing where we don't register an unknowable candidate if it's fully covered by another impl, I think.